### PR TITLE
Add support for Custom Preview Annotations

### PR DESCRIPTION
**Pull request by oas004**
## Add support for CustomMultiPreviewAnnotations in Showkase.

This is the third PR in a series to add support for Custom MultiPreview Annotations. Referencing as a followup on #259 and #255.

By completing this should solve #233 

## What do I mean by Custom MultiPreview Annotations?

This is an annotation in the form 

```kt
@Preview(name = "Custom Preview One First", group = "Custom Previews")
@Preview(name = "Custom Preview One Second", group = "Custom Previews")
annotation class CustomPreview
```

this can be used to annotate a composable function like you would do any other preview function in Compose like this:

```kt
@CustomPreview
@Composable
fun CustomAnnotationPreview() {
}
```

This can also be combined like this:

```kt
@Preview(name = "Custom Preview One First", group = "Custom Previews")
@Preview(name = "Custom Preview One Second", group = "Custom Previews")
annotation class CustomPreviewOne


@Preview(name = "Custom Preview Two First", group = "Custom Previews")
@Preview(name = "Custom Preview Two Second", group = "Custom Previews")
annotation class CustomPreviewTwo

@CustomPreviewOne
@CustomPreviewTwo
@Composable
fun CustomAnnotationPreviewCombined() {
}
```

They should be able to have the custom annotations and the use  of them in different modules in a project.

For more information, please see docs on tooling in compose [here](https://developer.android.com/jetpack/compose/tooling)

## Goal

The goal is to add support for this in Showkase. We would like showkase to act like the tooling in Android Studio does. That means that when Android Studio Preview would have generated a Preview, showkase should also generate a ShowkaseMetadata function.

### Sample App

In the sample app I have introduced `Shapes`. This is to showkase different shapes in Showkase.

Here I have made two internal (in same module) and one external (in submodule) annotations

Internals:

```kt
@Preview(name = "Square 200 by 200", group = "Shape", widthDp = 200, heightDp = 200)
@Preview(name = "Square 400 by 400", group = "Shape", widthDp = 400, heightDp = 400)
annotation class ShapePreview

@Preview(
    name = "Phone 200 by 200",
    group = "Shape",
    widthDp = 200,
    heightDp = 200,
    device = Devices.PHONE
)
@Preview(
    name = "Tablet 400 by 400",
    group = "Shape",
    widthDp = 400,
    heightDp = 400,
    device = Devices.TABLET
)
annotation class DevicePreview
```
Eksternal

```kt
@Preview(name = "Circle 150 by 150", group = "Shape", widthDp = 150, heightDp = 150)
@Preview(name = "Circle 200 by 200", group = "Shape", widthDp = 200, heightDp = 200)
annotation class CircleShapePreview
```

This is annotating different Composables in the `shapes` domain.  Here is an overview of the Composables that are annotated.

|  ElementName	|  Comined 	|   Scope	|
|---	|---	|---	|
|    CirclePreview	|   No	|   External	|
|   RoundRectanglePreview	|   Yes	|   Both	|
|   SquarePreview	|   No	|   Internal	|

The `RoundRectanglePreview` here is also annotated with a singular `Preview` annotation to check that the combination works. Since It is annotated with two annotations that has two `preview` annotations this should generate 2 + 2 + 1 previews in Showkase.

Please check out the app with the `useKsp=true `and click on the `shapes` tab to see the samples.

### Browser Test

For the UI tests I have currently made two custom annotations. One internal (same module) and one external(submodule)

Internal 

```kt
@Preview(
    name = "Custom Text Light",
    group = "Button",
    uiMode = Configuration.UI_MODE_NIGHT_NO,
)
@Preview(
    name = "Custom Text Dark",
    group = "Button",
    uiMode = Configuration.UI_MODE_NIGHT_YES,
)
annotation class CustomButtonPreview
```

External

```kt

@Preview(name = "CustomPreview one", group = "CustomExternalPreview")
@Preview(name = "CustomPreview two", group = "CustomExternalPreview")
annotation class CustomPreviewAnnotation
```

These are annotating different Composables. Here I have tried to combine them as well with each other and with singular ones.

Added tests for the different cases and checked that they are generating individual Composables in their respective group.
This should be the same as what is generated in the Sample application just with different names.

### Processor Test

In the processor tests, I have made tests that make sure all the correct metadata objects are created and that we are creating the files to store the custom annotations that the processor has registered. You can check these tests in `showcase-processor-testing` module.

## What is left

This is very much still a work in progress. I have tried to list up what I have left under to get an overview.
 
 <del>- [ ] Add docs about how to support your custom preview annotations.</del> - This should be done in a different PR that can be merged just before release so that it does not cause confusion.

 - [x] Make custom multi preview annotations work for annotations that are in the same module as the showkase preview.
 - [x] Make custom multi preview annotations work for annotations that are in a different module as the showkase preview and is imported into the file that has the showkase preview.
 - [x] Add processor tests for the different cases.
 
 <del>- [ ] Add processor testing submodule so that I can test that the annotation being in a different library module will generate a showkaseMetadata object. </del>   - Not possible because testing library does not support this
 
 - [x] Add Browser tests for the singular module case.
 - [x] Add Browser tests for the multi module case.
 - [x] Add processor test for combining custom annotations.
 - [x] Make combining multiple custom annotations generate multiple showkaseMetadata objects.
 - [x] Add Browser tests for combining multiple custom annotations.

### CODE REVIEW

File: showkase-processor-testing/build.gradle [hunk 1]
```diff
@@ -49,6 +49,10 @@ android {
     }
 }
 
+ksp {
+    arg("MultipreviewTypes", "com.airbnb.android.showkase.screenshot.testing.MyCustomAnnotation")
+}
+
```
Comment (oas004) [inline showkase-processor-testing/build.gradle:55]: I think I can remove this as it is provided through the test :)

File: showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt [hunk 2]
```diff
@@ -44,14 +33,30 @@ class ShowkaseProcessor @JvmOverloads constructor(
...
+            set.addAll(it)
+        }
+
+        return set
+    }
+
+    private fun getSupportedMultipreviewTypes(): Set<String>? {
+        return environment
+            .options["MultipreviewTypes"]
+            ?.split(",")?.map { it.replace(" ", "") }
```
Comment (oas004) [inline showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt:57]: This got somewhat ugly 🤔

File: showkase-processor/src/main/java/com/airbnb/android/showkase/processor/models/ShowkaseMetadata.kt [hunk 3]
```diff
@@ -298,6 +292,59 @@ internal fun getShowkaseMetadataFromPreview(
...
+    showkaseValidator: ShowkaseValidator,
+    annotationName: String,
+    roundEnv: XRoundEnv,
+): List<ShowkaseMetadata.Component?> {
+    val customAnnotation = element.requireAnnotationBySimpleName(annotationName)
+
+    val previewAnnotations = customAnnotation.map {
+//        it.typeElement.getAnnotations(PREVIEW_CLASS_NAME)
+        it.typeElement.requireAnnotationBySimpleName(PREVIEW_SIMPLE_NAME)
+    }.flatten()
```
Comment (oas004) [inline showkase-processor/src/main/java/com/airbnb/android/showkase/processor/models/ShowkaseMetadata.kt:307]: Here I ran into an issue when it comes to annotations that are in different modules. If they are in the same module, the typeElement has preview annotations. However, when they are in a different module, it never has any annotations. That list is always empty 🤔  Do you have any tips on how I can make that work @vinaygaba ? :) Is there some magic I can use here? 😅

Comment (oas004) [reply showkase-processor/src/main/java/com/airbnb/android/showkase/processor/models/ShowkaseMetadata.kt:307]: Maybe I can ask in the Kotlin KSP channel also if you don't know :)

Comment (vinaygaba) [reply showkase-processor/src/main/java/com/airbnb/android/showkase/processor/models/ShowkaseMetadata.kt:307]: If the user is using a custom annotation like `@MyCustomPreview`, they aren't using @Preview annotations in addition to that no? We have to implicitly get it based on the previews that are originally declared in  `@MyCustomPreview`

Comment (oas004) [reply showkase-processor/src/main/java/com/airbnb/android/showkase/processor/models/ShowkaseMetadata.kt:307]: Yeah, I was thinking, does that mean that I will need to ask the user to declare more than just the qualified name for the custom annotation in their gradle file? Like whatever information I would get from the `@Preview` annotation?

Because what I was trying to do here was to get the preview annotations that the `@MyCustomPreview` is annotated with. I am first getting the annotation from the `XRoundEnvironment` and then I thought since I knew the qualified name of the preview annotation I would be able to get that also from the `requireAnotationBySimpleName` or from the `XRoundEnvironment` somehow. 

IMO, this was somewhat confusing. Would be very nice with a call to discuss it :)

File: showkase-processor/src/main/java/com/airbnb/android/showkase/processor/models/ShowkaseMetadata.kt [hunk 4]
```diff
@@ -1,18 +1,12 @@
...
 
-import androidx.room.compiler.processing.XAnnotation
-import androidx.room.compiler.processing.XAnnotationBox
-import androidx.room.compiler.processing.XElement
-import androidx.room.compiler.processing.XFieldElement
-import androidx.room.compiler.processing.XMemberContainer
-import androidx.room.compiler.processing.XMethodElement
-import androidx.room.compiler.processing.XType
-import androidx.room.compiler.processing.XTypeElement
+import androidx.room.compiler.processing.*
```
Comment (oas004) [inline showkase-processor/src/main/java/com/airbnb/android/showkase/processor/models/ShowkaseMetadata.kt:3]: Will remove. Think I have some settings for wildcard imports that I have to change

File: showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt [hunk 5]
```diff
@@ -44,14 +33,52 @@ class ShowkaseProcessor @JvmOverloads constructor(
...
+
+        return set
+    }
+
+    private fun getSupportedMultipreviewTypes(): Set<String> {
+        val set = mutableSetOf<String>()
+        // This is to check if we have generated any types that we want to support.
+        set.addAll(
+            environment.getTypeElementsFromPackage(CODEGEN_PACKAGE_NAME)
+                .flatMap { it.getEnclosedElements() }.mapNotNull {
```
Comment (oas004) [inline showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt:68]: I have to clean this up

File: showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt [hunk 6]
```diff
@@ -114,6 +154,109 @@ class ShowkaseProcessor @JvmOverloads constructor(
...
+                        insideObject = commonMetadata.showkaseFunctionType.insideObject(),
+                        previewParameterName = customPreviewMetadata.previewName,
+                        previewParameterProviderType = previewParamMetadata?.second,
+                        showkaseGroup = getShowkaseGroup(
+                            customPreviewMetadata.previewGroup,
+                            commonMetadata.enclosingClass
+                        ),
+                        showkaseKDoc = commonMetadata.kDoc,
+                        enclosingClassName = commonMetadata.enclosingClassName,
+                        componentIndex = elementIndex + index, // Need this to be unique because of filtering algorithm. TOOD: Find a better way to do this
```
Comment (oas004) [inline showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt:246]: This I am unsure about. I kinda need this to be unique to remove the index from the name 🤔

File: showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt [hunk 7]
```diff
@@ -114,6 +154,109 @@ class ShowkaseProcessor @JvmOverloads constructor(
...
+                        insideWrapperClass = commonMetadata.showkaseFunctionType == ShowkaseFunctionType.INSIDE_OBJECT,
+                        showkaseHeightDp = if (customPreviewMetadata.showkaseHeight == -1) null else customPreviewMetadata.showkaseHeight,
+                        showkaseWidthDp = if (customPreviewMetadata.showkaseWidth == -1) null else customPreviewMetadata.showkaseWidth,
+                    )
+
+                }
+        }.flatten().mapNotNull { it }.toSet()
+    }
+
+    private fun String.getCustomAnnotationSimpleName(): String {
```
Comment (oas004) [inline showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt:256]: I can remove this now I think

File: showkase-processor/src/main/java/com/airbnb/android/showkase/processor/logging/ShowkaseValidator.kt [hunk 8]
```diff
@@ -297,7 +296,10 @@ internal class ShowkaseValidator {
             }
         }
     }
-
+    internal fun validateCustomAnnotation(customAnnotationElements: Set<XElement>, environment: XProcessingEnv) {
+        if (customAnnotationElements.isEmpty()) return
+        // TODO: Write validator
```
Comment (oas004) [inline showkase-processor/src/main/java/com/airbnb/android/showkase/processor/logging/ShowkaseValidator.kt:301]: I am unsure if I actually need this 🤔 Will look into this

File: sample-submodule/build.gradle [hunk 9]
```diff
@@ -3,14 +3,14 @@ plugins {
...
-    kapt {
-        correctErrorTypes = true
-    }
-}
+//} else {
+//    apply plugin: 'kotlin-kapt'
+//    kapt {
+//        correctErrorTypes = true
+//    }
+//}
```
Comment (oas004) [inline sample-submodule/build.gradle:13]: Will revert these changes. It was just to make it easier to test locally. Sorry about this

File: sample/src/main/java/com/airbnb/android/showkasesample/CustomPreview.kt [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package com.airbnb.android.showkasesample
+
+import androidx.compose.ui.tooling.preview.Preview
+
+//@Preview(name = "Basic Chip Red First", group = "Chips")
+//@Preview(name = "Basic Chip Red Second", group = "Chips")
+//public annotation class CustomPreview
```
Comment (oas004) [inline sample/src/main/java/com/airbnb/android/showkasesample/CustomPreview.kt:7]: Think I should have this to test if the single modular case also works.

File: showkase-browser-testing/build.gradle [hunk 11]
```diff
@@ -10,6 +10,10 @@ if (project.hasProperty('useKsp')) {
     }
 }
 
+ksp {
+    arg("MultipreviewTypes", "com.airbnb.android.showkase_browser_testing.CustomButtonPreview")
+}
+
```
Comment (oas004) [inline showkase-browser-testing/build.gradle:16]: Don't need this. Will remove it

File: sample/build.gradle [hunk 12]
```diff
@@ -70,6 +70,10 @@ if (project.hasProperty("useKsp")) {
 }
 
 
+ksp {
+    arg("MultipreviewTypes", "com.airbnb.android.showkasesample.CustomPreview")
+}
```
Comment (oas004) [inline sample/build.gradle:75]: I think I can remove this now. Might need it for KAPT, but works otherwise with KSP :)

File: showkase-processor/src/main/java/com/airbnb/android/showkase/processor/writer/ShowkaseBrowserWriter.kt [hunk 13]
```diff
@@ -72,6 +69,47 @@ internal class ShowkaseBrowserWriter(private val environment: XProcessingEnv) {
...
+                ).build()
+        }
+
+        fileBuilder.addType(
+            TypeSpec.classBuilder(generatedClassName).addFunctions(functions).build()
+        )
+        try {
+            fileBuilder.build().writeTo(environment.filer, mode = XFiler.Mode.Aggregating)
+        } catch (fileExists: FileAlreadyExistsException) {
+            // Then the annotation already exists in a file and we don't want to write to it.
```
Comment (oas004) [inline showkase-processor/src/main/java/com/airbnb/android/showkase/processor/writer/ShowkaseBrowserWriter.kt:108]: Is this correct to assume at this stage?

File: showkase-screenshot-testing/build.gradle [hunk 14]
```diff
@@ -69,4 +69,15 @@ dependencies {
...
+    // Compose
+    implementation deps.compose.activityCompose
+    implementation deps.compose.composeRuntime
+    implementation deps.compose.composeNavigation
+    implementation deps.compose.core
+    implementation deps.compose.foundation
+    implementation deps.compose.tooling
+    implementation deps.compose.layout
+    implementation deps.compose.material
+    androidTestImplementation deps.compose.uiTest
```
Comment (oas004) [inline showkase-screenshot-testing/build.gradle:82]: Think I can remove these changes 🤔

File: .github/workflows/android.yml [hunk 15]
```diff
@@ -56,7 +56,7 @@ jobs:
           api-level: ${{ matrix.api-level }}
           target: ${{ matrix.target }}
           arch: x86_64
-          script: ./gradlew connectedCheck --no-daemon --stacktrace -PuseKsp=false
+          script: ./gradlew connectedCheck --no-daemon --stacktrace
```
Comment (vinaygaba) [inline .github/workflows/android.yml:59]: Why was this needed?

Didn't we figure out a way to avoid running something on kapt in tests directly as opposed to skipping it here?

Comment (oas004) [reply .github/workflows/android.yml:59]: Yeah I thought so as well. After watching the CI build that failed it looked like it was building KSP tests at that stage. Removing this flag made it run the KAPT version at least. I was suspecting that it checked if we had that flag, not that it was set to true. I can remove this change ofc, but then I am unsure if it will pass CI.

File: showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt [hunk 16]
```diff
@@ -44,14 +46,53 @@ class ShowkaseProcessor @JvmOverloads constructor(
...
+        val set = mutableSetOf(
+            ShowkaseComposable::class.java.name,
+            PREVIEW_CLASS_NAME,
+            ShowkaseColor::class.java.name,
+            ShowkaseTypography::class.java.name,
+            ShowkaseRoot::class.java.name,
+            ShowkaseScreenshot::class.java.name,
+        )
+
+        environment.getTypeElementsFromPackage(CODEGEN_PACKAGE_NAME)
```
Comment (vinaygaba) [inline showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt:61]: hmm I'm not sure if we should be doing this here 🤔

Comment (oas004) [reply showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt:61]: hmm, yeah that is a good point actually 🤔 Wouldn't that always be the case for incremental builds though if I understand this correctly?

Comment (vinaygaba) [reply showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt:61]: I think what you need here is the custom multi preview annotation name instead of the custom annotation that we've created for codegen purposes

File: showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt [hunk 17]
```diff
@@ -121,6 +184,97 @@ class ShowkaseProcessor @JvmOverloads constructor(
...
+        supportedTypes.addAll(getSupportedMultipreviewTypes())
+
+        return supportedTypes.map { supportedAnnotation ->
+            val elementsAnnotated = roundEnvironment.getElementsAnnotatedWith(supportedAnnotation)
+            return@map elementsAnnotated.map elementScope@{ element ->
+                if (showkaseValidator.checkElementIsAnnotationClass(element)) {
+                    // Here we write to metadata to aggregate custom annotation data.
+                    // In this case, it would be a custom annotation that is annotated
+                    // with something custom. Eg. CombinedPreview.
+                    ShowkaseBrowserWriter(environment).writeCustomAnnotationElementToMetadata(
```
Comment (vinaygaba) [inline showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt:206]: We are writing in two places. Need to double check if we need to do this or doing it just once is sufficient

File: showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt (hunk 16 above)
Comment (oas004) [reply showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt]: Removed this and added the multi preview name from the options :)

File: showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt (hunk 17 above)
Comment (oas004) [reply showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt]: Removed writing to file here. Seems like you where correct. We do not need it here :) Thanks for pointing that out, great catch! :)

File: .github/workflows/android.yml (hunk 15 above)
Comment (oas004) [reply .github/workflows/android.yml:95]: I think that we actually need this change from testing this locally. I think it is since we check `project.hasProperty('useKsp')` and that checks that we have the property, not that it is true.

File: showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt [hunk 18]
```diff
@@ -44,14 +46,43 @@ class ShowkaseProcessor @JvmOverloads constructor(
...
+        )
+
+        environment
+            .options["MultiPreviewTypes"]
+            ?.split(",")?.map { it.replace(" ", "") }
+            ?.toSet()?.let { set.addAll(it) }
+        return set
+    }
+
+    private fun getSupportedMultipreviewTypes(): Set<String> {
```
Comment (vinaygaba) [inline showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt:68]: move these two functions below where it's actually being used, that way it's easier to follow in code for future code reviews

File: showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt [hunk 19]
```diff
@@ -73,10 +104,25 @@ class ShowkaseProcessor @JvmOverloads constructor(
...
-    private fun processComponentAnnotation(roundEnvironment: XRoundEnv): Set<ShowkaseMetadata.Component> {
+    private fun processComponentAnnotation(
+        roundEnvironment: XRoundEnv,
+    ): Set<ShowkaseMetadata.Component> {
         val showkaseComposablesMetadata = processShowkaseAnnotation(roundEnvironment)
         val previewComposablesMetadata = processPreviewAnnotation(roundEnvironment)
-        return (showkaseComposablesMetadata + previewComposablesMetadata)
+
+        // This is for getting custom annotations from class path
+        val customAnnotationMetadataFromClassPath =
```
Comment (vinaygaba) [inline showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt:114]: Why are we picking from the classpath here? Why not do it at the same place where we do it for `@Preview` and `@ShowkaseComposable` annotations

Comment (oas004) [reply showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt:114]: I think the reason I have to split this up into two methods, is that the `processCustomAnnotation` method only uses the supported annotations type qualified name and was meant for KAPT to work. But the `processCustomAnnotationFromClassPath` was meant to get the stored annotation metadata from each round and based on that check if we had any qualified names we could search for. The reason we have to split them up is that we are using some of the other properties from the stored metadata that we store on each round I believe.  Maybe I can remove the `processCustomAnnotation` and merge that into the `processCustomAnnotationFromClassPath` but I think that might risk it becoming very messy? And I kinda don't know how to handle that we get preview name and such from the stored metadata and not from the supported compile option name thingy 🤔

File: showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt (hunk 18 above)
Comment (oas004) [reply showkase-processor/src/main/java/com/airbnb/android/showkase/processor/ShowkaseProcessor.kt]: Good idea :) Moved them in [50bf2f5](https://github.com/airbnb/Showkase/pull/263/commits/50bf2f500ce613997694172deeea1e0459d7e14b)

**Comment by polivmi1** [comment]
Is this still planned in near future? Is any help needed?

**Comment by oas004** [comment]
> Is this still planned in near future? Is any help needed?

With this PR it should work for KAPT. We had to do some changes for KSP. I think I will have to make another PR because there is to much conflicts and we have to change the approach a bit after some of the changes that has been merged. I have not gotten around to it quite yet because of other things. If you want to help and check this out, you are more than welcome to :)

**Comment by anhanh11001** [comment]
Hi @oas004 , thanks for the work!

My team’s work is dependent on this custom preview annotation in order to support Paparazzi screenshot testing. Can I ask if there are estimated timeline on having this released (only KSP is already great to us)?

I’d love to help if I can do anything but I’m really new to annotation processing programming so maybe guides are needed

**Comment by oas004** [comment]
> Hi @oas004 , thanks for the work!
> 
> My team’s work is dependent on this custom preview annotation in order to support Paparazzi screenshot testing. Can I ask if there are estimated timeline on having this released (only KSP is already great to us)?
> 
> I’d love to help if I can do anything but I’m really new to annotation processing programming so maybe guides are needed

Hey @anhanh11001. Sorry for blocking you. Had a bit to much on my plate at work, so this has been a bit down prioritised. I will try my best to get some work in on this this week. Since we have released #255, it should at least not throw at compile time, it will just be bypassed from Showkase point of view.

I think the stuff in this PR should be enough for it to work with KSP at least. It has to be changed along with this #284

**Comment by oas004** [comment]
Closing this in favour of #303
